drivers/SmartThings/philips-hue/src/handlers/attribute_emitters.lua

```diff
@@ -125,30 +112,175 @@ local function _emit_light_events_inner(light_device, light_repr)
+      )
+    )
+  end
+
+  local button_idx_map = button_device:get_field(Fields.BUTTON_INDEX_MAP)
+  if not button_idx_map then
+    log.error(
+      string.format(
+        "Button ID to Button Index map lost, " ..
+        "cannot find componenet to emit attribute event on for button [%s]",
+        (button_device and button_device.lable) or "unknown button"
+      )
+    )
+  end
+
+  local idx = button_idx_map[button_info.id] or 1
+  local component_idx
+  if idx == 1 then
+    component_idx = "main"
+  else
+    component_idx = string.format("button%s", idx)
+  end
+
+  local button_report = button_info.button.button_report or { event = "" }
+
+  if button_report.event == "long_press" and not button_device:get_field("button_held") then
+    button_device:set_field("button_held", true)
+    button_device.profile.components[component_idx]:emit_event(
+      capabilities.button.button.held({state_change = true})
+    )
+  end
+
+  if button_report.event == "long_release" and button_device:get_field("button_held") then
+    button_device:set_field("button_held", false)
+  end
+
+  if button_report.event == "short_release" and not button_device:get_field("button_held") then
+    button_device.profile.components[component_idx]:emit_event(
+      capabilities.button.button.pushed({state_change = true})
+    )
+  end
+end
+
+function AttributeEmitters.emit_contact_sensor_attribute_events(sensor_device, sensor_info)
+  if sensor_device == nil or (sensor_device and sensor_device.id == nil) then
+    log.warn("Tried to emit attribute events for a device that has been deleted")
+    return
+  end
+
+  if sensor_info.power_state then
+    log.debug(true, "emit power")
+    sensor_device:emit_event(capabilities.battery.battery(st_utils.clamp_value(sensor_info.power_state.battery_level, 0, 100)))
+  end
+
+  if sensor_info.tamper_reports then
+    log.debug(true, "emit tamper")
+    local num_reports = #sensor_info.tamper_reports
+    local not_tampered = 0
+    for _, tamper in ipairs(sensor_info.tamper_reports) do
+      if tamper.state == "not_tampered" then
+        not_tampered = not_tampered + 1
+      end
+    end
+
+    if not_tampered == num_reports then
+      sensor_device:emit_event(capabilities.tamperAlert.tamper.clear())
+    else
+      sensor_device:emit_event(capabilities.tamperAlert.tamper.detected())
+    end
+  end
+
+  if sensor_info.contact_report then
+    log.debug(true, "emit contact")
+    if sensor_info.contact_report.state == "contact" then
+      sensor_device:emit_event(capabilities.contactSensor.contact.closed())
+    else
+      sensor_device:emit_event(capabilities.contactSensor.contact.open())
+    end
+  end
+end
+
+function AttributeEmitters.emit_motion_sensor_attribute_events(sensor_device, sensor_info)
+  if sensor_device == nil or (sensor_device and sensor_device.id == nil) then
+    log.warn("Tried to emit attribute events for a device that has been deleted")
+    return
+  end
+
+  if sensor_info.power_state then
```

Comment:
Would it make sense to pull this whole check into a 

`handle_power_state_if_needed` 

function that would do this check and emit if needed?  Or is the structure of the info not consistent enough across the devices?

Reply:
I needed to finish all the devices to answer this question, and the answer is that yes it can be generalized.

The entire `sensor` class of devices (of which Signify includes Buttons) can be generalized quite a bit. But I didn't wanna back myself in to a crappy design by doing a DRY refactor too early.